### DISCUSSION TITLE
scaling socketIO with multiple workers on Heroku

### DISCUSSION THREAD

**Discussion by WeekiCode** [miguelgrinberg/Flask-SocketIO#1777] [Q&A]
Hi,

Flask/socketIO does not work in Heroku with gunicorn workers >1.  Application works fine on Heroku with worker=1.  I get various socket failure with workers=2 (see log below [engineio_logger=True]).  You have given various answers that say to use multiple copies of gunicorn, instead of multiple workers.  Heroku support says they do not support multiple gunicorns in one Dyno.  Not sure I believe them.   Do you know of people using socketIO in Heroku that have scaled?  Is there another way of running multiple flask/socketIO servers in order to scale?
[heroku socketIO log.txt](https://github.com/miguelgrinberg/Flask-SocketIO/files/7966921/heroku.socketIO.log.txt)

**Comment by bluthen** [comment]
This is only an answer to the last question. I've never used Heroku.
The workaround could be very application dependent. I've made a couple applications were we use multiple socketio servers.

- We use sticky sessions and had a application were we didn't need to share information between other clients
- We had a application were the sharing of information among other clients in only in small groups, so you can use sticky sessions based on a custom cookie were all the people in the small group get the same cookie and hit the same socket io server.

Otherwise you need to have some way for the servers to share information as well as sticky sessions. A message queue with: Redis, kafka, or something else.
https://flask-socketio.readthedocs.io/en/latest/deployment.html#using-multiple-workers

**Reply by WeekiCode** [reply]
In Heroku sticky sessions are called "affinity" and enabled with the command: heroku features:enable http-session-affinity --app xxxxxxxx

I enabled it and it ran fine with a single worker.  When I added a second worker, I started to get a lot of:  Invalid session 50EQeR4s0LS2v9-SAAAA.  I got this message for more than one ID. From the app's perspective, the items that depend on socketIO did not work at all.

Log file is attached...
[after stick and 2 workers change.txt](https://github.com/miguelgrinberg/Flask-SocketIO/files/7989468/after.stick.and.2.workers.change.txt)


Any ideas?

**Comment by miguelgrinberg** [comment]
I suggest you read the documentation wrt using multiple workers. The link @bluthen has provided is correct.

Running multiple gunicorn processes in a single Heroku dyno is not something they want you to do. I have never tried, but it may work if you create a bash script that launches all the Gunicorns as background processes, and then put this bash script in your Procfile. In any case, are you sure you need multiple workers? If you are using gevent or eventlet, then a single worker is often enough, especially for constrained platforms such as Heroku.

**Reply by WeekiCode** [reply]
I am really uncertain of what I need to scale.  I plan on launching the application in the next few weeks, and am trying to anticipate solutions to scaling problems that will arise in the near future.  Given the various reads (your book, your youtube tutorial, and others things I have read), I assumed that eventlet or gevent needed help from the likes of gunicorn.  Heroku Dyno's come in different sizes, so the limitation would be the entity that gives me parallelism.  How far have you seen gevent and eventlet go in scaling?

**Reply by miguelgrinberg** [reply]
This isn't a question that can be answered universally. If you want to know how your application scales, then you have to test it. All applications are different.

**Reply by WeekiCode** [reply]
Clearly, each application is different.  I am new to web applications, so I could  have this wrong.  Having the scale be governed by eventlet means that eventlet must be very thin and the top layer of the application must be very thin.  One green thread can block all the others.  
 At scale, the sheer number of green threads is the bottleneck.  Isn't that a limiting factor, even if they are very thin?  So at some level, to plan going forward having some idea of the kinds of scale eventelt can reach would be useful, even in a theoretical way.  

You are right, I will need to do some testing. I was hoping for some rules of thumb as a guide for when to get paranoid about scale.  Searching Google yields anecdotes, but so far, no good data.

**Reply by miguelgrinberg** [reply]
I'm not sure I follow your logic. Eventlet can handle thousands, even tens of thousands of running greenlets in a single worker process. The bottleneck are the resources consumed by your application.

**Reply by WeekiCode** [reply]
If I understand it correctly, eventlet runs one green thread at a time.  Adding CPU's does not make any difference to scaling. Scaling with eventlet means  that the green thread is so small as to not be the bottleneck.  I am assuming that the real work is put off to Celery, as you suggest in multiple write-ups.  That means, for a given CPU, eventlet is asymptotic towards some number as you can only make the green thread so small.  The obvious question is how relevant that is to planning for application scaling.

For example, when a request comes in, it takes my app about 2msec to get the relevant data about the request through SQLAlchemy and Postgres.  This would argue that 500 threads per second would max out the performance.  Given that queueing theory says CPU loads usually should be more like 50% of the channel bandwidth, it says I can do some 250 threads/sec.  This all implies that the database should not be in the green thread.  So I am left wondering what a reasonable number to assume I can achieve in a flask or socketIO  CPU times on the green thread side of the equation.  Is 500useconds per green thread transaction a reasonable goal?

I have no idea if I am worried about something that I should not be, or if indeed this is something I need to plan for.   As I re-read what I wrote above, I am left wondering if I am overly concerned.

Thoughts?

**Reply by WeekiCode** [reply]
I meant to write  "is 50 usec per transaction in the green thread a reasonable goal?"

**Reply by miguelgrinberg** [reply]
My answer is the same as before. There are no universal numbers, it all depends on your application, so the only way you'll find answers is by stress-testing your application to find what its limits are. There is really no other way.

> If I understand it correctly, eventlet runs one green thread at a time.

This is not related to eventlet. Python runs one thread at a time. If you have N CPUs, the only way to utilize all of them concurrently is to run N copies of your Python process, or else use a specialized package such as `multiprocessing` to create a child process for each CPU.

> I am assuming that the real work is put off to Celery

This is highly dependent on the application. Not all applications need Celery.

> I have no idea if I am worried about something that I should not be

I think your worries are reasonable. You should find actual answers through testing to understand if the worries are justified or not.

**Reply by WeekiCode** [reply]
Is it safe to assume the Gunicorn provides that ability to add N CPU's to the equation?

**Reply by miguelgrinberg** [reply]
If you are asking as a general question, then yes, Gunicorn can start multiple worker processes. But you already know that this is insufficient for Socket.IO, since this is what started this discussion.

**Reply by WeekiCode** [reply]
I meant, staring multiple gunicorn's (1 worker each), but we have discussed that in relation to Heroku.  I am not sure how to scale with Heroku beyond what one eventlet can do.  

Thanks very much for your patience and time!

**Reply by miguelgrinberg** [reply]
In Heroku the unit is the dyno. To scale, you add dynos.

**Comment by WeekiCode** [comment]
thanks @bluthen..... I will try it tomorrow and report back
